Progress bar docs (#2688)
carbon-design-system/carbon-website: src/pages/components/progress-bar/usage.mdx

information about the progression in the optional helper text.

### Anatomy

<Row>
<Column colLg={8}>

![Progress bar anatomy](images/progress-bar-usage-6.png)

</Column>
</Row>

1. **Label:** The text describes the process that the progress bar refers to.
   The label can be visually hidden, but must be defined in order to be
   accessible.
2. **Track:** The static area that the bar indicator moves on top of and acts as
   a fixed visual reference of what the total length and duration of the process
   could be.
3. **Bar indicator:** Indicates how much the process has progressed.
4. **Helper text (optional):** Assistive text to show additional information
   about the process taking place. For determinate, the information is usually
   values of number or percentage. Sometimes, a fraction or ratio may be shown.
   These values are not applicable when a progress bar is indeterminate.

### Sizing

#### Height

The progress bar is offered in two different sizes—default (8px) and small
(4px). The default progress bar height is typically used when there is large
amounts of space on a page. The small progress bar height is commonly used when

> space is restricted and can be placed within cards, data table, or side panels.

Reviewer: ```suggestion
space is restricted and can be placed within cards, data tables, or side panels.
```

<Row>
<Column colLg={8}>

![Progress bar sizes](images/progress-bar-usage-7.png)

</Column>
</Row>

#### Width

The width of a progress bar can be customized appropriately for its context. The
minimum width of a progress bar is 48px and keep its width to a maximum of six
columns when possible. If the progress bar width is too long it can reduce
readability. Don't increase the progress bar length to fill the entire width of
a window or application.

<DoDontRow>

  <DoDont caption="Do keep a width within six columns.">

![Do keep a width within six columns.](images/progress-bar-usage-8.png)

  </DoDont>
  <DoDont type="dont" caption="Don't exceed a width of six columns.">

![Don't exceed a width of six columns.](images/progress-bar-usage-9.png)

  </DoDont>
</DoDontRow>

### Alignment

Text alignment depends on the context and area given. As long as the text is
placed close to the progress bar, it’s up to designers to provide the best
judgement that fits the context.

<Row>
<Column colLg={12}>

![Progress bar alignments](images/progress-bar-usage-10.png)

> <Caption>A few common examples of text alignmnet for progress bars</Caption>

Reviewer: ```suggestion
<Caption>A few common examples of text alignment for progress bars</Caption>
```

</Column>
</Row>

#### Text placement

Labels or helper text should always be placed above, below, or beside the

> progress bar for readibility. Avoid placing text inside or centering text with

Reviewer: ```suggestion
progress bar for readability. Avoid placing text inside or centering text with
```

  </DoDont>
  <DoDont type="dont" caption="Don’t embed text into the progress bar.">

![Don’t embed text into the progres bar](images/progress-bar-usage-13.png)

  </DoDont>
</DoDontRow>

<DoDontRow>
  <DoDont caption="Do place text close to the progress bar.">

![Do place text close to the progress bar](images/progress-bar-usage-14.png)

  </DoDont>
  <DoDont type="dont" caption="Don't place text far away from the progress bar.">

![Don't place text far away from the progress bar](images/progress-bar-usage-15.png)

  </DoDont>
</DoDontRow>

#### Internationalization (RTL)

For RTL (right-to-left) languages, the layout of the progress bar is mirrored
for both determinate and indeterminate options. The label is right-aligned, the
value is left-aligned, and the fill progresses from right to left. Keep in mind
that the placement of the percent sign differs depending on the locale.

<Row>
<Column colLg={6}>

![Progress bar RTL](images/progress-bar-usage-11.gif)

</Column>
</Row>

## Content

### Main elements

#### Labels

- Labels inform users what information the progress bar is processing.
- Keep the label short and concise by limiting it to a single line of text.

> - Labels should never change as the progres bar is loading.

Reviewer: ```suggestion
- Labels should never change as the progress bar is loading.
```

#### Helper text

- Helper text is used to show additional information about the process taking
  place.
- Common usecases of helper text for a progress bar can either be a generic
  phrase, for example “Fetching assets…”, or it can indicate an exact amount of
  something to be completed, for example "42/256 items".
- When indicating a loading percentage, the percentage number should count up as
  the bar progresses.

### Further guidance

For further content guidance, see Carbon's
[content guidelines]([https://www.carbondesignsystem.com/guidelines/content/overview/]).

## Universal behaviors

### States

There are three states for the progress bar: active, success, and error. These

> states also apply on both indeterminate and determinate progress bars. After the

Reviewer: ```suggestion
states also apply to both indeterminate and determinate progress bars. After the
```

process completes successfully or unsuccesfully, a progress bar can either
remain persistent as confirmation or validation, or it can be automatically
dismissed depending on what is most suitable for the use case.

<Row>
<Column colLg={8}>

![Progress bar states.](images/progress-bar-usage-16.png)

</Column>
</Row>

#### Active

The active loading state indicates that the action is still in progress.

> #### Successful

Reviewer: I would change this state label to read as "Success".

The success loading state indicates that the action completed successfully.

#### Error

The error loading state indicates that the action did not successfully complete.
If an error occurs, an inline notification or error handling within the form
should appear. As soon as a process fails, the current visual progress remains
visible but the indicator bar's progress stops in place and spans full width.

<Row>
<Column colLg={8}>

![The example shows the progress bar with error remains persistent as confirmation inside the image tile](images/progress-bar-usage-17.gif)

<Caption>

>   The example shows the progress bar with error remains persistent as

Reviewer: ```suggestion
  The example shows the progress bar with an error remains persistent as
```

  confirmation inside the image tile.
</Caption>

</Column>
</Row>

<Row>
<Column colLg={8}>

> ![The example shows thep progress bar is automatically dismissed once it’s done uploading](images/progress-bar-usage-18.gif)

Reviewer: ```suggestion
![The example shows the progress bar is automatically dismissed once it’s done uploading](images/progress-bar-usage-18.gif)
```

<Caption>

>   The example shows thep progress bar is automatically dismissed once it’s done

Reviewer: ```suggestion
  The example shows the progress bar is automatically dismissed once it’s done
```

  uploading.
</Caption>

</Column>
</Row>

### Interactions

#### Part of a page

While the progress bar is processing information on part of a page, you can
still interact with other elements on the page. When the progress bar is

> finished processing it can either remain persistant on the page for confirmation

Reviewer: ```suggestion
finished processing it can either remain persistent on the page for confirmation
```

> or automatically dismiss itself depending on the usecase.

Reviewer: ```suggestion
or automatically dismiss itself depending on the use case.
```

- When the progress is determined by user actions such as tutorials completed,
  or storage space, rather than system actions
- If the process takes less than 5 seconds to load, use
  [loading spinners](/components/loading/usage) instead

## Live demo

## Variants

A progress bar can either be determinate or indeterminate. The determinate
variant is visualized as a track on which a bar fills up from left to right
representing the progress. The indeterminate variant is visualized as a track on
which a bar is moving along in constant speed, repeating over time.

| Variant                        | Purpose                                                                                                                                                                                                                                                    |
| ------------------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --- |

> | [Determinate](Determinate)     | Use determinate progress bars for a specified amount of time or clear information about the progression — they convey exact and qualitative information about the progression.                                                                             |     |

Reviewer 2: Links aren't working 

```suggestion
| [Determinate](/components/progress-bar/usage#determinate-progress-bar)     | Use determinate progress bars for a specified amount of time or clear information about the progression — they convey exact and qualitative information about the progression.                                                                             |     |
```

> | [Indeterminate](Indeterminate) | Use indeterminate progress bars for an unspecified completion of time or unclear information about the progression — they convey that users’ request, action, or data is being processed at the moment without indicating how long the activity will take. |

Reviewer 2: Link isn't working
```suggestion
| [Indeterminate](/components/progress-bar/usage#indeterminate-progress-bar) | Use indeterminate progress bars for an unspecified completion of time or unclear information about the progression — they convey that users’ request, action, or data is being processed at the moment without indicating how long the activity will take. |
```

<Row>
<Column colLg='4'>

![Determinate progress bars fill from 0 to 100%.](images/progress-bar-usage-2.gif)

<Caption>
  Determinate progress bars fill from 0 to 100%.
</Caption>
</Column>

<Column colLg='4'>

![Indeterminate progress bars move along a fixed track continually until the process is complete.](images/progress-bar-usage-3.gif)

<Caption>
  Indeterminate progress bars move along a fixed track continually until the process is complete.
</Caption>
</Column>
</Row>

> ## Determinate progress bar

Reviewer 3: No, there is no image for this

process completes successfully or unsuccesfully, a progress bar can either
remain persistent as confirmation or validation, or it can be automatically
dismissed depending on what is most suitable for the use case.

<Row>
<Column colLg={8}>

![Progress bar states](images/progress-bar-usage-16.png)

</Column>
</Row>

#### Active

The active loading state indicates that the action is still in progress.

#### Success

The success loading state indicates that the action completed successfully.

#### Error

The error loading state indicates that the action did not successfully complete.
If an error occurs, an inline notification or error handling within the form
should appear. As soon as a process fails, the current visual progress remains
visible but the indicator bar's progress stops in place and spans full width.

<Row>
<Column colLg='8'>

<Video src={localVideo3} poster={localPoster3} />

<Caption>
  The example shows the progress bar with error remains persistent as
  confirmation inside the image tile.
</Caption>

</Column>
</Row>

<Row>
<Column colLg='8'>

<Video src={localVideo4} poster={localPoster4} />

<Caption>

>   The example shows thep progress bar is automatically dismissed once it’s done

Reviewer 2: ```suggestion
  The example shows the progress bar is automatically dismissed once it’s done
```

---
title: Progress bar
description:
  A progress bar provides feedback about the duration and progression of a
  process.
tabs: ['Usage', 'Style']
---

<PageDescription>

A progress bar provides feedback about the duration and progression of a
process.

</PageDescription>

<AnchorLinks>

<AnchorLink>Overview</AnchorLink>
<AnchorLink>Live demo</AnchorLink>
<AnchorLink>Variants</AnchorLink>
<AnchorLink>Formatting</AnchorLink>
<AnchorLink>Content</AnchorLink>
<AnchorLink>Universal behaviors</AnchorLink>
<AnchorLink>Modifiers</AnchorLink>
<AnchorLink>Accessibility</AnchorLink>
<AnchorLink>Related</AnchorLink>
<AnchorLink>References</AnchorLink>
<AnchorLink>Feedback</AnchorLink>

</AnchorLinks>

## Overview

A progress bar indicates that the user's request has been received and the
application is making progress toward completing the requested action. Progress
bars inform users about the status of ongoing processes, the estimated time of

> how long a process will take, or a request is being excuted.

Reviewer 4: ```suggestion
how long a process will take, or if a request is being executed.
```

<Row>
<Column colLg={12}>

![Progress bar hero image](images/progress-bar-usage-1.png)

</Column>
</Row>

#### When to use

> - If a process can take a considerable or unknown amount of time
> - When the process can be described with quantitative information, for example
>   as a percentage
> - To visually show the progression of a system operation such as downloading,
>   uploading, loading data, submitting a form, or saving updates
> - To convey that data is being requested, transferred or processed

Reviewer 4: This might be a purposeful style decision, so correct me if I'm wrong, but I think any list items of sentence length should have periods.

```suggestion
- If a process can take a considerable or unknown amount of time.
- When the process can be described with quantitative information, for example
  as a percentage.
- To visually show the progression of a system operation such as downloading,
  uploading, loading data, submitting a form, or saving updates.
- To convey that data is being requested, transferred or processed.
```

There are other spots in the document where this needs to be updated as well. I won't flag them since it's quite a few and I'm not 100% sure if this is a purposeful decision to omit periods.

Reviewer 3: In this case, we want to remove the periods after these sentences because they are not full sentences.

#### When not to use

- When the progress bar will be replaced by a minor user interface element once
  loaded, use [skeleton states](/patterns/loading) instead
- When manual user actions are required to progress, use a
  [progress indicator](/components/progress-indicator/usage) instead
- When the progress is determined by user actions such as tutorials completed,
  or storage space, rather than system actions

> - If the process takes less than 10 seconds to load, use

Reviewer 2: I just noticed some of you gif examples take less than 10 seconds but feel correct. I wonder if 10 seconds is too long?

  [loading spinners](/components/loading/usage) instead

#### Loading, inline loading, or progress bar?

> - If an activity displays a loading spinners on one screen, it should not show

Reviewer 4: ```suggestion
- If an activity displays a loading spinner on one screen, it should not show
```

  the progress bar elsewhere in the design.
- Both progress bars and circles can show either determinate or indeterminate
  progress. The given area should help determine if a progress bar or circle is
  best for that space. Progress bars are preferred in vertically narrow areas
  (tables, cards, etc.) while progress circles are ideal for full page loading
  or in very small areas.

## Live demo

## Variants

A progress bar can either be determinate or indeterminate. The determinate
variant is visualized as a track on which a bar fills up from left to right
representing the progress. The indeterminate variant is visualized as a track on
which a bar is moving along in constant speed, repeating over time.

> When it is neccesary to show the percentage and the descriptive text, the

Reviewer 3: both you and Taylor said that, so I'd just remove it. It makes sense, we don't say about that anywhere on our site.

developers will have to calculate the percentage on their side with the value:

> `${(value/max*100).toFixed(2)}% (${value} of ${max} items)`

Reviewer 4: It might be better to put this in storybook docs rather than on the website? Seems like a piece of minutia not directly related to component guidance.

| Variant                        | Purpose                                                                                                                                                                                                                                                    |
| ------------------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --- |
| [Determinate](Determinate)     | Use determinate progress bars for a specified amount of time or clear information about the progression — they convey exact and qualitative information about the progression.                                                                             |     |
| [Indeterminate](Indeterminate) | Use indeterminate progress bars for an unspecified completion of time or unclear information about the progression — they convey that users’ request, action, or data is being processed at the moment without indicating how long the activity will take. |

<Row>
<Column colLg={4}>

> ![Determinate progress bars fill from 0 to 100%.](images/progress-bar-usage-2.gif)

Reviewer 2: Gifs need to use the [gif component](https://gatsby-theme-carbon.vercel.app/components/GifPlayer) so that it has a pause button on them.  

```
<Column` colLg='4'>
<GifPlayer>

![IBM Cloud Pictograms](/images/IBM_Cloud_Pictograms.gif) // must be gif

![IBM Cloud Pictograms](/images/IBM_Cloud_Pictograms.jpg) // must be static
image

</GifPlayer>
```

Reviewer 4: @thyhmdo @laurenmrice this is why the gif is disappearing on pause - a static image needs to be given.

<Caption>Determinate progress bars fill from 0 to 100%.</Caption>

</Column>

<Column colLg={4}>

![Indeterminate progress bars move along a fixed track continually until the process is complete.](images/progress-bar-usage-3.gif)

<Caption>
  Indeterminate progress bars move along a fixed track continually until the
  process is complete.
</Caption>

</Column>
</Row>

## Determinate progress bar

Determinate progress bars are used when the loading progress can be tracked and
translated into a percentage indicator which fills from 0% to 100% and never

> decrease in value. The entire animation moves from the left to the right – with

Reviewer 4: ```suggestion
decreases in value. The entire animation moves from the left to the right – with
```

no back and forth motion.

Use a determinate progress bar when progress can be calculated against a
specific goal. For example, when downloading a file of a known size.

<Row>
<Column colLg={8}>

![Progress bar uploading](images/progress-bar-usage-4.gif)

</Column>
</Row>

> ## Indeterminate progress bar
> 
> Once triggered, the loading bar will begin by moving along in constant speed,

Reviewer 4: It feels like there should be a quick explanation of when Indeterminate is used. Just to set the context - particularly because these section headers are bookmarkable and the link stored or sent as a reference.

Just a super rough idea here:

```suggestion
## Indeterminate progress bar

Indeterminate progress bars are used when loading progress is unknown or likely to take a long time. These can be replaced by a determinate progress bar if a completion estimate for a process becomes known or calculated.

Once triggered, the loading bar will begin by moving along in constant speed,
```

repeating over time while data is being gathered during the first stages of the
load. This bar then changes to a determinate indicator (finishing from left to
right) when enough information is gathered to end the load animation. The final
determinate stage of the load is used to visually indicate the completion of the
task to the user. In this scenario, a percentage indicator would not be present.

<Row>
<Column colLg={8}>

> ![Progress bar uploading](images/progress-bar-usage-5.gif)

Reviewer 2: This example isn't changing to a determinate indicator in second stage like the paragraph says.

provide progress estimation in certain cases. If progress data is not available,
an indeterminate progress bar can be used.

As a more concrete set of examples, there are generally three different major
progressions of a system operation: downloading, uploading, and local
processing.

#### Downloading

For downloading actions it's usually possible to get a value of progress and
update it in real time for a determinate progress bar. Depending on the
information reported from a data source, time remaining could be estimated based
on how much has been downloaded so far over a time period.

#### Uploading

Upload actions may be unable to report real time progress status depending on
the data source and API configuration. However, always strive to provide users
the determinate value whenever it's possible to inform them about the process.

#### Data processing

Data processing refers to manipulations of data to produce meaningful
information. This can include currency exchange, file generation, data
reformatting, or similar operations. Depending on the operation, time estimated
or cycles to complete processing could be used to populate a determinate
progress bar.

## Formatting

> Determinate and indeterminate progress bars both have a label and an optional

Reviewer 2: This information seems out of place in the general formatting section. I'd move it either to include it in the anatomy or content section.

helper text. However, the inderterminate can’t provide the exact, quantiative
information about the progression in the optional helper text.

### Anatomy

<Row>
<Column colLg={8}>

![Progress bar anatomy](images/progress-bar-usage-6.png)

</Column>
</Row>

1. **Label:** The text describes the process that the progress bar refers to.

>    Label can be visually hidden, but requires to be defined to be accessible.

Reviewer 4: ```suggestion
   The label can be visually hidden, but must be defined in order to be accessible.
```

>    Label can be visually hidden, but requires to be defined to be accessible.

Reviewer 2: ```suggestion
   Label can be visually hidden, but is required to be defined to be accessible.
```

2. **Track:** The static area that the bar indicator moves on top of and acts as
   a fixed visual reference of what the total length and duration of the process
   could be.
3. **Bar indicator:** Indicates how much the process has progressed.
4. **Helper text (optional):** Assistive text to show additional information
   about the process taking place. For determinate, the information is usually
   values of number or percentage. Sometimes, a fraction or ratio may be shown.
   These values are not applicable when a progress bar is indeterminate.

### Sizing

#### Height

The progress bar is offered in two different sizes—default (8px) and small
(4px). The default progress bar height is typically used when there is large
amounts of space on a page. The small progress bar height is commonly used when
space is restricted and can be placed within cards, data table, or side panels.

<Row>
<Column colLg={8}>

![Progress bar sizes](images/progress-bar-usage-7.png)

</Column>
</Row>

#### Width

> The width of a progress bar can be customized appropriately for its context. The
> minimum width of a progress bar is 48px and keep its width to a maximum of six
> columns when possible. If the progress bar width is too long it can reduce
> readability. Don't increase the progress bar length to fill the entire width of
> a window or application.

Reviewer 4: Brief suggestion

```suggestion
The width of a progress bar can be customized appropriately for its context. The
minimum width of a progress bar is 48px and it should be kept to a maximum of six
columns wide when possible. If the progress bar width is too long it can reduce
readability. Don't increase the progress bar length to fill the entire width of
a window or application.
```

<DoDontRow>

  <DoDont caption="Do keep a width within six columns.">

![Do keep a width within six columns.](images/progress-bar-usage-8.png)

  </DoDont>
  <DoDont type="dont" caption="Don't exceed a width of six columns.">

![Don't exceed a width of six columns.](images/progress-bar-usage-9.png)

  </DoDont>
</DoDontRow>

### Alignment

Text alignment depends on the context and area given. As long as the text is
placed close to the progress bar, it’s up to designers to provide the best
judgement that fits the context.

<Row>
<Column colLg={12}>

![Progress bar alignments](images/progress-bar-usage-10.png)

<Caption>A few common examples of text alignmnet for progress bars</Caption>

</Column>
</Row>

#### Internationalization (RTL)

For RTL (right-to-left) languages, the layout of the progress bar is mirrored
for both determinate and indeterminate options. The label is right-aligned, the
value is left-aligned, and the fill progresses from right to left. Keep in mind
that the placement of the percent sign differs depending on the locale.

<Row>
<Column colLg={6}>

![Progress bar RTL](images/progress-bar-usage-11.gif)

</Column>
</Row>

> #### Text placement

Reviewer 2: I would move this section above section "Internationalization" section because it related more to that paragraph under alignment. They seem to cover similar related topics

  place.
- Common usecases of helper text for a progress bar can either be a generic
  phrase, for example “Fetching assets…”, or it can indicate an exact amount of
  something to be completed, for example "42/256 items".
- When indicating a loading percentage, the percentage number should count up as
  the bar progresses.

### Further guidance

For further content guidance, see Carbon's
[content guidelines]([https://www.carbondesignsystem.com/guidelines/content/overview/]).

## Universal behaviors

### States

There are three states for the progress bar: active, success, and error. These
states also apply on both indeterminate and determinate progress bars. After the
process completes successfully or unsuccesfully, a progress bar can either
remain persistent as confirmation or validation, or it can be automatically
dismissed depending on what is most suitable for the use case.

<Row>
<Column colLg={8}>

![Progress bar states.](images/progress-bar-usage-16.png)

</Column>
</Row>

#### Active

The active loading state indicates that the action is still in progress.

#### Successful

The success loading state indicates that the action completed successfully.

#### Error

> The error loading state indicates that the action did not successfully complete.

Reviewer 2: Sentence is repeated here.
```suggestion

```

> The error loading state indicates that the action did not successfully complete.
> The error loading state indicates that the action did not successfully complete.

Reviewer 4: Duplicate sentence here
```suggestion
The error loading state indicates that the action did not successfully complete.
```

If an error occurs, an inline notification or error handling within the form

> should appear. As soon as a process fails, the current visual progress remains

Reviewer 2: Add a little more detail here to help with understanding like: 
"When the process fails in determinate bars, the current progress remains visible but indicator bar's progress stops in place." 

I also think indeterminate wouldn't stop incrementally on error but span full width like in the example below. I would add a bit about that state.

progress bar.

<Row>
<Column colLg={8}>

![Progress bar inside a container](images/progress-bar-usage-20.png)

</Column>
</Row>

## Modifiers

#### Transitions

An indeterminate bar can change to a determinate indicator (finishing from left
to right) when enough information is gathered to end the load animation.

<Row>
<Column colLg={8}>

![An indeterminate bar can change to a determinate indicator](images/progress-bar-usage-21.gif)

</Column>
</Row>

## Accessibility

- The loading component is implemented using the ARIA 1.1 progressbar pattern.
  It outputs an ARIA `role="progressbar"` and uses `aria-valuemin`,
  `aria-valuemax`, and `aria-valuenow` to convey the loaded percentage to screen
  reader users.

- If the progress bar shows the loading of a specific part or region of the
  applicaition, set `aria-busy="true"` and `aria-describedly="[progressbar-id]"`
  on the related element to ensure assistive technologies can correctly convey
  this relation to the user.

- Only hide the label when it is absolute clear to the user which process the
  progress bar represents. Note that you are still required to pass an

>   appropriate label which will be ready by screen readers.

Reviewer 4: ```suggestion
  appropriate label which will be read by screen readers.
```

## Related

#### Components

- [Loading spinners](/components/loading/usage)
- [Inline loading](/components/inline-loading/usage)
- [Progress indicator](/components/progress-indicator/usage)

#### Components

- [Loading: skeleton states](/patterns/loading/#skeleton-states)

#### Loading, inline loading, or progress bar?

- If an activity displays a loading spinners on one screen, it should not show
  the progress bar elsewhere in the design.
- Both progress bars and circles can show either determinate or indeterminate
  progress. The given area should help determine if a progress bar or circle is
  best for that space. Progress bars are preferred in vertically narrow areas
  (tables, cards, etc.) while progress circles are ideal for full page loading

>   or in very small areas.

Reviewer 4: Should this duplicate part be removed? Feels like it's nearly identical to what is written up above

Reviewer 3: @aagonzales @laurenmrice think we talked briefly about this, saying that it's good to remind people about this. Is it too redundant?
